From d7fbca20aa00e98f54b78deb1d3c145f61d37540 Mon Sep 17 00:00:00 2001
From: Davide Cavalca <dcavalca@fb.com>
Date: Mon, 27 Nov 2017 14:08:43 -0800
Subject: [PATCH] network: add knob to optionally keep interfaces up during
 shutdown

---
 rc.d/init.d/network | 9 +++++++++
 sysconfig.txt       | 5 +++++
 2 files changed, 14 insertions(+)

diff --git a/rc.d/init.d/network b/rc.d/init.d/network
index 2b8c191..ae5956f 100755
--- a/rc.d/init.d/network
+++ b/rc.d/init.d/network
@@ -170,6 +170,15 @@ case "$1" in
 	fi
 
         systemctl show --property=RequiredBy -- -.mount | grep -q 'remote-fs.target' && exit 1
+
+  # Don't shut the network down when shutting down the system if configured
+  # as such in sysconfig
+  if [ "$IFDOWN_ON_SHUTDOWN" = no ]; then
+    if systemctl is-system-running | grep -q 'stopping'; then
+      net_log "system is shutting down, leaving interfaces up as requested"
+      exit 1
+    fi
+  fi
   
 	vlaninterfaces=""
 	vpninterfaces=""
diff --git a/sysconfig.txt b/sysconfig.txt
index e34ca1b..7a9c3d1 100644
--- a/sysconfig.txt
+++ b/sysconfig.txt
@@ -178,6 +178,11 @@ Generic options:
     network has spanning tree running and must wait for STP convergence.
     Default: 0 (no delay)
 
+  IFDOWN_ON_SHUTDOWN=yes|no
+    If yes, do bring interfaces down during system shutdown. If no, leave them
+    in their current state (this is only supported on hosts using systemd).
+    Default: yes (bring interfaces down)
+
 
   IPV6FORWARDING=yes|no
     Enable or disable global forwarding of incoming IPv6 packets 
-- 
2.11.0

