From ab87d03e9a2e4b831558a75ec512ea0e2f4c5f62 Mon Sep 17 00:00:00 2001
From: Davide Cavalca <dcavalca@fb.com>
Date: Fri, 2 Sep 2016 11:07:50 -0700
Subject: [PATCH] Fix arch mirror selection

---
 mkosi | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mkosi b/mkosi
index 3872610..e88c158 100755
--- a/mkosi
+++ b/mkosi
@@ -1275,7 +1275,7 @@ def load_args():
             args.mirror = "http://httpredir.debian.org/debian"
         elif args.distribution == Distribution.ubuntu:
             args.mirror = "http://archive.ubuntu.com/ubuntu"
-        elif Distribution.arch:
+        elif args.distribution == Distribution.arch:
             args.mirror = "https://mirrors.kernel.org/archlinux"
 
     if args.bootable:
-- 
2.8.0.rc2

