From 8201a901337f160ff242dff0574ef9f3025d103d Mon Sep 17 00:00:00 2001
From: Roman Gushchin <guro@fb.com>
Date: Thu, 23 Aug 2018 10:45:26 -0700
Subject: [PATCH 1/4] libbpf.h: add BPF_JMP_A macro

---
 src/shared/linux/libbpf.h | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/src/shared/linux/libbpf.h b/src/shared/linux/libbpf.h
index ad131fcb1..391eee5a4 100644
--- a/src/shared/linux/libbpf.h
+++ b/src/shared/linux/libbpf.h
@@ -174,6 +174,16 @@ struct bpf_insn;
                 .off   = OFF,					\
                 .imm   = IMM })
 
+/* Unconditional jumps */
+
+#define BPF_JMP_A(OFF)						\
+        ((struct bpf_insn) {					\
+                .code  = BPF_JMP | BPF_JA,			\
+                .dst_reg = 0,					\
+                .src_reg = 0,					\
+                .off   = OFF,					\
+                .imm   = 0 })
+
 /* Raw code statement block */
 
 #define BPF_RAW_INSN(CODE, DST, SRC, OFF, IMM)			\
-- 
2.17.1

