From 4c2fa3b381824e1a2d622e8f5cf0ada68ed365cc Mon Sep 17 00:00:00 2001
From: Davide Cavalca <dcavalca@fb.com>
Date: Tue, 26 Feb 2019 16:09:40 -0500
Subject: [PATCH] fs-util: add missing linux/falloc.h include

---
 src/basic/fs-util.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/basic/fs-util.c b/src/basic/fs-util.c
index f25bf2cbbcb..3ff86157973 100644
--- a/src/basic/fs-util.c
+++ b/src/basic/fs-util.c
@@ -6,6 +6,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <sys/stat.h>
+#include <linux/falloc.h>
 #include <linux/magic.h>
 #include <time.h>
 #include <unistd.h>
