diff --git a/src/basic/def.h b/src/basic/def.h
index 970654a..c61d910 100644
--- a/src/basic/def.h
+++ b/src/basic/def.h
@@ -57,7 +57,7 @@
 #define CONF_PATHS_STRV(n)                      \
         STRV_MAKE(CONF_PATHS(n))
 
-#define HIGH_RLIMIT_MEMLOCK (1024ULL*1024ULL*64ULL)
+#define HIGH_RLIMIT_MEMLOCK (1024ULL*1024ULL*512ULL)
 
 #define PLYMOUTH_SOCKET {                                       \
                 .un.sun_family = AF_UNIX,                       \
