From d407285a625b69bc72219b514ee3c8bc98b7c263 Mon Sep 17 00:00:00 2001
From: Davide Cavalca <dcavalca@fb.com>
Date: Wed, 13 Sep 2017 16:35:00 -0700
Subject: [PATCH] basic: ensure O_TMPFILE is always defined

---
 src/basic/missing.h | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/src/basic/missing.h b/src/basic/missing.h
index 7830a4f415..0269066427 100644
--- a/src/basic/missing.h
+++ b/src/basic/missing.h
@@ -606,14 +606,13 @@ struct input_mask {
 #else
 #define __O_TMPFILE     020000000
 #endif
+#endif
 
 /* a horrid kludge trying to make sure that this will fail on old kernels */
 #ifndef O_TMPFILE
 #define O_TMPFILE (__O_TMPFILE | O_DIRECTORY)
 #endif
 
-#endif
-
 #if !HAVE_DECL_LO_FLAGS_PARTSCAN
 #define LO_FLAGS_PARTSCAN 8
 #endif
