diff --git a/test/meson.build b/test/meson.build
index 36d9df7..d3c0435 100644
--- a/test/meson.build
+++ b/test/meson.build
@@ -264,17 +264,6 @@ custom_target(
         output : 'sys',
         build_by_default : want_tests != 'false')
 
-if perl.found()
-        udev_test_pl = find_program('udev-test.pl')
-        if want_tests != 'false'
-                test('udev-test',
-                     udev_test_pl,
-                     timeout : 180)
-        endif
-else
-        message('Skipping udev-test because perl is not available')
-endif
-
 if conf.get('ENABLE_HWDB') == 1
         hwdb_test_sh = find_program('hwdb-test.sh')
         if want_tests != 'false'
