diff --git a/src/test/meson.build b/src/test/meson.build
index 7da7e3a..f82f855 100644
--- a/src/test/meson.build
+++ b/src/test/meson.build
@@ -550,18 +550,6 @@ tests += [
           libmount,
           libblkid]],
 
-        [['src/test/test-execute.c',
-          'src/test/test-helper.c'],
-         [libcore,
-          libshared],
-         [threads,
-          librt,
-          libseccomp,
-          libselinux,
-          libmount,
-          libblkid],
-         '', 'timeout=360'],
-
         [['src/test/test-siphash24.c'],
          [],
          []],
