commit c59f5fae6f369e6f6a97e71d2b252821c031016d
Author: Davide Cavalca <dcavalca@fb.com>
Date:   Thu Feb 22 12:58:34 2018 -0800

    FB: workaround issue 8194

diff --git a/src/nspawn/nspawn-setuid.c b/src/nspawn/nspawn-setuid.c
index b08bcd9..bbe0d40 100644
--- a/src/nspawn/nspawn-setuid.c
+++ b/src/nspawn/nspawn-setuid.c
@@ -34,7 +34,7 @@
 #include "util.h"
 
 static int spawn_getent(const char *database, const char *key, pid_t *rpid) {
-        int pipe_fds[2], r;
+        int pipe_fds[2];
         pid_t pid;
 
         assert(database);
@@ -44,10 +44,10 @@ static int spawn_getent(const char *database, const char *key, pid_t *rpid) {
         if (pipe2(pipe_fds, O_CLOEXEC) < 0)
                 return log_error_errno(errno, "Failed to allocate pipe: %m");
 
-        r = safe_fork("(getent)", FORK_RESET_SIGNALS|FORK_DEATHSIG|FORK_LOG, &pid);
-        if (r < 0)
-                return r;
-        if (r == 0) {
+        pid = fork();
+        if (pid < 0)
+                return log_error_errno(errno, "Failed to fork getent child: %m");
+        else if (pid == 0) {
                 int nullfd;
                 char *empty_env = NULL;
 
