From a23b95a82129b460eb35e4303e424338fe56e5ed Mon Sep 17 00:00:00 2001
From: Davide Cavalca <dcavalca@fb.com>
Date: Mon, 18 Sep 2017 15:27:01 -0700
Subject: [PATCH] fix libraries versions

---
 meson.build | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/meson.build b/meson.build
index eb515a0..e309ab7 100644
--- a/meson.build
+++ b/meson.build
@@ -81,7 +81,7 @@ libsystemd_daemon = shared_library(
         'systemd-daemon',
         join_paths([systemd_src, 'libsystemd/sd-daemon/sd-daemon.c']),
         libsystemd_daemon_src,
-        version : '0.12.0',
+        version : '0.0.12',
         include_directories : includes,
         c_args : cargs,
         link_args : [
@@ -98,7 +98,7 @@ libsystemd_id128 = shared_library(
         'systemd-id128',
         id128_sources,
         libsystemd_id128_src,
-        version : '0.28.0',
+        version : '0.0.28',
         include_directories : includes,
         c_args : cargs,
         link_args : ['-shared',
@@ -113,7 +113,7 @@ libsystemd_login = shared_library(
         'systemd-login',
         join_paths([systemd_src, 'libsystemd/sd-login/sd-login.c']),
         libsystemd_login_src,
-        version : '9.3.9',
+        version : '0.9.3',
         include_directories : includes,
         c_args : cargs,
         link_args : ['-shared',
@@ -131,7 +131,7 @@ libsystemd_journal = shared_library(
         join_paths([systemd_src, 'libsystemd/sd-event/sd-event.c']),
         join_paths([systemd_src, 'libsystemd/sd-daemon/sd-daemon.c']),
         [libsystemd_journal_src],
-        version : '11.5.11',
+        version : '0.11.5',
         include_directories : includes,
         c_args : cargs,
         link_args : ['-shared',
-- 
2.9.5

